controllers/clusteroperator_controller.go

```diff
@@ -144,17 +188,48 @@ func (r *CloudOperatorReconciler) syncStatus(ctx context.Context, co *configv1.C
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *CloudOperatorReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&configv1.ClusterOperator{}, builder.WithPredicates(predicate.Funcs{
-			CreateFunc:  func(e event.CreateEvent) bool { return clusterOperatorFilter(e.Object) },
-			UpdateFunc:  func(e event.UpdateEvent) bool { return clusterOperatorFilter(e.ObjectNew) },
-			GenericFunc: func(e event.GenericEvent) bool { return clusterOperatorFilter(e.Object) },
-			DeleteFunc:  func(e event.DeleteEvent) bool { return clusterOperatorFilter(e.Object) },
-		})).
-		Complete(r)
+	watcher, err := NewObjectWatcher(WatcherOptions{
+		Cache:  mgr.GetCache(),
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return err
+	}
+	r.watcher = watcher
+
+	build := ctrl.NewControllerManagedBy(mgr).
+		For(
+			&configv1.ClusterOperator{},
+			builder.WithPredicates(clusterOperatorPredicates()),
+		).
+		Watches(
+			&source.Kind{Type: r.Object()},
```

Comment:
`r.Object` is very not obvious here what this is, where does this come from, the platform owner? 🤔  Might be best to just have the type in here so it's obvious
```suggestion
			&source.Kind{Type: &configv1.Infrastructure{}},
```

Reply:
I'll show a POC for CAPI resources later. It is meant to do that.

Reply:
@JoelSpeed Here is a POC how it will work for CAPI cluster with minimal set of changes: https://github.com/openshift/cluster-cloud-controller-manager-operator/commit/d844b87d308df3f24474f42193063895885e4c2f

Reply:
Are we planning on integrating with CAPI? I had assumed this would be an OpenShift specific project and therefore we would always rely on the Infrastructure object. I think unless we have an immediate need for supporting CAPI we should keep this simple and perhaps only add this abstraction when we actually need it

Reply:
I'm addressing the comment on the proposal. https://github.com/openshift/enhancements/pull/463#discussion_r561994942

Reply:
Let's leave this implementation on it's own branch and we can return to the interface if it is needed. Ok with watching `Infrastructure` resources only.